pkgs/development/compilers/llvm/16/llvm/gnu-install-dirs.patch

```diff
@@ -49,8 +49,10 @@ index 057431208322..56f0dcb258da 100644
 -    set(_install_rpath "${LLVM_LIBRARY_OUTPUT_INTDIR}" "${CMAKE_INSTALL_PREFIX}/lib${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
 +    set(_install_rpath "${LLVM_LIBRARY_OUTPUT_INTDIR}" "${CMAKE_INSTALL_FULL_LIBDIR}${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
    elseif(UNIX)
--    set(_install_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
-+    set(_install_rpath "\$ORIGIN/../${CMAKE_INSTALL_LIBDIR}${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
+-    set(_build_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
+-    set(_install_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}")
++    set(_build_rpath "\$ORIGIN/../${CMAKE_INSTALL_LIBDIR}${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
++    set(_install_rpath "\$ORIGIN/../${CMAKE_INSTALL_LIBDIR}${LLVM_LIBDIR_SUFFIX}")
```

Comment:
Absolutely. I'd rather avoid a repeat of messing up llvm again, so take as long as you need!

Reply:
Apologies for the delay.

----

I ended up just adding `${extra_libdir}` back to `_install_rpath`: https://github.com/rrbutani/nixpkgs/commit/666014fe2c731540675c94747431da8b468f8df7; this unbreaks the `x86_64-linux` build. (I haven't tested darwin/aarch64-linux yet but I think it should be fine; if you cherry-pick that commit we can have ofborg give it a go)

I feel a little uneasy adding a patch that has no route to being upstreamed but unfortunately this seems like the least-worst solution; our other options (updating LLVM rdeps to add to `CMAKE_INSTALL_RPATH` either manually or via a setup hook or applying patches to rdeps) all seem too disruptive or brittle.

Despite the discussion on the [differential](https://reviews.llvm.org/D146918) this does _seem_ like a pretty common usecase (depending on a prebuilt/standalone LLVM). For example, other distros do provide LLVM dev packages with `AddLLVM.cmake` and friends (i.e. Ubuntu's [`llvm-15-dev`](https://launchpad.net/ubuntu/+source/llvm-toolchain-15)). It's not apparent to me if/how this patch doesn't break those flows; is it just FHS (assuming the install dir for LLVM and the dependent ends up being some subdir in `/usr/lib`)?

I'd feel better about this if some of the other LLVM nixpkgs maintainers took a look; perhaps I'm missing something obvious.

cc: @alyssais @RaitoBezarius @Ericson2314 @sternenseemann

Reply:
The patch seems very fair to me, but I do not have more ideas about the ramifications of this whole situation. Though I don't feel strongly about it.

Reply:
> I was a little disappointed (per discussion) to see this break and tried but failed to argue that the old behaviour should be kept.

Thanks for trying anyways 🙂

> Having a patch to fix this does seem unfortunate if it could be handled by updating `CMAKE_INSTALL_RPATH`. Which seems a big if.

I think ^ would work but I believe we'd have to add LLVM's lib dir to `CMAKE_INSTALL_RPATH` in the build of every LLVM dependent (which seems untenable — without something like a setup hook that does this we'd break out-of-nixpkgs use cases I think).

Does ^ match your understanding? Perhaps I'm missing something.

Reply:
That seems right to me, at least. All of my use cases are very much out of nixpkgs. I suppose we could just tell people that this is what they'd need to add but that seems rude too. Right now I'm leaning toward the patch.

Reply:
Actually, the more I think about it, I feel like the downsides are pretty low. We're already putting in a patch for gnu-install-dirs no matter what, which already broke going from 16.0.1 to 16.0.5 (will soon update to .6), so seems like adding one more change in there isn't the worst thing in the world.

Reply:
> Still leaning toward the patch, but I suppose switching to the CMAKE variable could be pursued later once it's clearer there's no additional breakage.

I agree with @rrbutani -- the issue is that if LLVM's cmake infrastructure is not patched before it is installed, anything which points itself at the installed LLVM during configure time, when built, will not have the required rpath. The installed LLVM cmake infrastructure ideally should express how to link to the installation at runtime for packages depending on it. Otherwise, users are left expressing how to link it at configure time: If a user wants to develop a package pointing at the LLVM installation, they would also need to set the `CMAKE_INSTALL_RPATH` (since the usual cmake package discovery infrastructure would not be enough to set this).

TL;DR, I can't see a route which isn't patching this at the moment, or having a hook (which would require builds of other packages depending on LLVM to work within the nixpkgs infrastructure to pick up the rpath, I think).

The other ideal situation from nixpkgs point of view (and cmake projects depending on LLVM generally) would be for the upstream patch to be reverted -- but then infrastructure building packages without installing them would need changing to specify that they are doing this (and the build rpath should be dropped).

Reply:
> @rrbutani to confirm, the change, were we not to modify `CMAKE_INSTALL_RPATH`, would just be to change the `gnu-install-dirs` patch's revised from ...
> (snipped)

Yup!

I think it'd be good to leave a comment explaining the situation (like [this](https://github.com/willcohen/nixpkgs/compare/emscripten-wasm2c...rrbutani:nixpkgs:llvm-16-bump#diff-9173dc5067b4d545691792b0062164d5e2297e91ac168f9cbadabeb3527493a6R52-R75)), preferably in the actual `AddLLVM.cmake` file in case anyone ends up looking through it to debug stuff.

> We're already putting in a patch for gnu-install-dirs no matter what, which already broke going from 16.0.1 to 16.0.5 (will soon update to .6), so seems like adding one more change in there isn't the worst thing in the world.

I think the difference is that the rest of the `gnu-install-dirs` patch (IIRC — away from a computer so can't double check right now) is in the process of being upstreamed or is at least upstream-able. Agreed that it's not a huge deal though.

Reply:
> TL;DR, I can't see a route which isn't patching this at the moment, or having a hook (which would require builds of other packages depending on LLVM to work within the nixpkgs infrastructure to pick up the rpath, I think).

Good point re: having to use the nixpkgs infrastructure; there probably actually _are_ users that use LLVM from nixpkgs in devShells and what not (where the env vars set by setup hooks may not actually be picked up by the actual build tools).

> The other ideal situation from nixpkgs point of view (and cmake projects depending on LLVM generally) would be for the upstream patch to be reverted -- but then infrastructure building packages without installing them would need changing to specify that they are doing this (and the build rpath should be dropped).

IIRC it was mentioned that upstream would be amenable to accepting a build knob for this; maybe we can look into that if keeping this patch up to date proves difficult.